### PR TITLE
backport #21933 to 3.0 branch 

### PR DESCRIPTION
**Pull request by Liu-ErMeng**
##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

relate to #23018

### CODE REVIEW

**Comment by Liu-ErMeng** [comment]
@nhorman ，is the `fips-provider-validation`  test failed caused by newly added code snippet in `evp_test.c`?

File: test/evp_test.c
Comment (t8m) [inline test/evp_test.c:699]: I think this should be `fips_provider_version_gt(libctx, 3, 0, 12)`

File: test/evp_test.c
Comment (t8m) [inline test/evp_test.c:839]: You need to add:
```
        EVP_CIPHER_CTX_free(ctx);
        ctx = ctx_base;
```
here

File: test/evp_test.c
Comment (t8m) [inline test/evp_test.c:841]: Add an else case where you do:
```
        EVP_CIPHER_CTX_free(ctx_base);
        ctx_base = NULL;
```

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by nhorman** [comment]
Will look as soon as.i get to my desk

**Comment by Liu-ErMeng** [comment]
> > @nhorman , help! is the `fips-provider-validation` test failed caused by newly added code snippet in `evp_test.c`?
> 
> Yes the CI is relevant.

thx, I will fix these.

**Comment by Liu-ErMeng** [comment]
> > @nhorman , help! is the `fips-provider-validation` test failed caused by newly added code snippet in `evp_test.c`?
> 
> Yes the CI is relevant.

Hi @t8m , I have fix these issues. Please review again.

**Review by t8m — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Review by nhorman — APPROVED** [review]

**Comment by t8m** [comment]
This requires another review from @openssl/committers as @nhorman is the original author.

File: providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c
Comment (mattcaswell) [inline providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c:378]: It is strange the all the other dupctx functions check if `ctx` is NULL first - but this one doesn't. Actually it makes no difference since `OPENSSL_memdup` checks for NULL anyway so it ends up doing the right thing.

Anyway I realise this is a backport, so this is just a "FYI" comment

**Review by mattcaswell — APPROVED** [review]

File: providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c
Comment (Liu-ErMeng) [reply providers/implementations/ciphers/cipher_aes_cbc_hmac_sha.c:378]: yes, this is a backport.

**Comment by t8m** [comment]
Merged to the 3.0 branch. Thank you for your contribution.

**Comment by adrien-n** [comment]
I am seeing test failures on 3.0.13 which are related to this merge request.

For the testsuite to pass, I have to revert the following commits:
- 58a6aa0c9fe6abad996f45c6b452983035db7105
- 879a853a1dc968fb010e5bf17d2e8888acc70742
- 0398bc20080de037a8433fe81cfdef3ba0ec9d4c

**Comment by rsbeckerca** [comment]
`shlibtest` needs to be disabled during the back-port.

**Comment by t8m** [comment]
> I am seeing test failures on 3.0.13 which are related to this merge request.
> 
> For the testsuite to pass, I have to revert the following commits:
> 
>     * [58a6aa0](https://github.com/openssl/openssl/commit/58a6aa0c9fe6abad996f45c6b452983035db7105)
> 
>     * [879a853](https://github.com/openssl/openssl/commit/879a853a1dc968fb010e5bf17d2e8888acc70742)
> 
>     * [0398bc2](https://github.com/openssl/openssl/commit/0398bc20080de037a8433fe81cfdef3ba0ec9d4c)

As we do not see the failures in our CI could you please open a  new issue and report the failures that you see?

**Comment by t8m** [comment]
@adrien-n ^

**Comment by adrien-n** [comment]
Sorry for not answering earlier; GH needs a reaction that amounts to "ack, thanks for your answer, I'm currently looking into it and it might take me some time".

I also didn't answer earlier because I had a weird feeling following the comments here. It turns out that the issue happens only when we also patch in a backport of at least one of commits 8a75952ba829784f5cb499c4883a5729c226e38c or 949108dd73de321fb93c8d81b846a2a1d015a9fd . Right now I see several possible causes:
- the CPU support patches,
- the backport of these patches,
- the combination of these patches and the patches I mentioned earlier,
- something else.

@t8m I'll therefore open a bug when I can tell which one of these possibilities it is in order to make a proper one; this is taking a while as the builds are LTO (including tests) (and, checking whether the builds should be LTO is already scheduled in the coming weeks).

**Pull request closed without merging**
